Fix #30971: Handle requisites correctly for empty SLS files

dwoz opened the pull request:
Empty SLS files were not being tracked properly for requisite checking, causing 'requisite not found' errors. This fix ensures empty SLS files are processed and tracked correctly for requisite dependencies.

- Added _processed_sls_files set to track processed SLS files
- Track SLS files in compile_high_data even if they produce no chunks
- Check _processed_sls_files when validating SLS requisites
- Handle empty SLS requisites in call_chunk execution
- Track SLS files in render_state and get_highstate
- Added test case to test_require.py to verify the fix

Fixes #30971

dwoz committed:
Fix #30971: Handle requisites correctly for empty SLS files

Empty SLS files were not being tracked properly for requisite checking,
causing 'requisite not found' errors. This fix ensures empty SLS files
are processed and tracked correctly for requisite dependencies.

- Added _processed_sls_files set to track processed SLS files
- Track SLS files in compile_high_data even if they produce no chunks
- Check _processed_sls_files when validating SLS requisites
- Handle empty SLS requisites in call_chunk execution
- Track SLS files in render_state and get_highstate
- Added test case to test_require.py to verify the fix

bdrx312 commented on salt/state.py:
```diff
@@ -1231,8 +1233,6 @@ def _run_check_cmd(self, low_data):
         cmd_opts = {}
         if "shell" in self.opts["grains"]:
             cmd_opts["shell"] = self.opts["grains"].get("shell")
-        if isinstance(low_data["check_cmd"], str):
-            low_data["check_cmd"] = [low_data["check_cmd"]]
```
This looks like it was inadvertently removed

dwoz replied:
Restored in 4d116c4.

twangboy reviewed (changes requested):
Changelog

dwoz committed:
Address review: simplify empty SLS requisite handling, add changelog

- Keep import from salt.utils.datastructures (odict is deprecated)
- Restore str-to-list conversion for check_cmd that was inadvertently removed
- Remove unused local processed_sls_files in compile_high_data; track on
  self._processed_sls_files instead
- Replace the duplicated empty-SLS handling in call_chunk with a single
  post-loop check in check_requisite and the lost-requisite builder
- Drop redundant tracking in render_highstate and include processing; the
  render_state path already records the SLS in _processed_sls_files
- Add changelog entry

dwoz commented on salt/state.py:
```diff
@@ -52,7 +52,7 @@
 from salt.serializers.msgpack import serialize as msgpack_serialize
 from salt.template import compile_template, compile_template_str
-from salt.utils.datastructures import DefaultOrderedDict, HashableOrderedDict
+from salt.utils.odict import DefaultOrderedDict, HashableOrderedDict
```
Reverted in 4d116c4 - salt.utils.odict is deprecated, kept import from salt.utils.datastructures.

dwoz commented on salt/state.py:
```diff
@@ -1718,9 +1718,16 @@ def compile_high_data(self, high, orchestration_jid=None):
+        # Track all SLS files that were processed, even if they produced no chunks
+        # This is needed to handle SLS files that produce no output but are still
+        # required by other states (Issue #30971)
+        processed_sls_files = set()
```
Right, replaced with self._processed_sls_files in 4d116c4.
